Signpost API Energy Improvements

adkinsjd opened the pull request:
These are the signpost API changes that I found necessary to run useful energy adaptation and attribution code on signpost.

A summary of the current changes:
- changing the energy info struct to be simpler
- adding energy reporting functionality (and two new structs)
  - This is primarily for the radio to report other modules using its power
- Fixing some bugs in the energy query code
- exposing addr_to_key and addr_to_mod_num functions for the controller to use in its application when dealing with energy
- adding an energy_duty_cycle function that modules can call

What needs to be done moving forward to get this merged:
 - Finalize energy API (should it be more expressive/time period based?)
   - If the API changes the signpost_energy_policy and signpost_controller_send_bytes_app/main.c should change to reflect that update
 - Finalize energy reporting API
   - Same as above - any changes here need to propagate down to energy_policy, controller app and radio app that use the reporting API
 - Pull Duty cycling and energy implementations out of specific controller app and into a separate file that can be used by multiple controller apps
  - After pulling these out they should be pushed into the other controller apps as well
 - Other general cleanups - the controller app is a bit of a mess right now.

adkinsjd committed:
pulled out just signpost_api changes into another branch

adkinsjd committed:
cleaning up functions for merge

adkinsjd committed:
added energy_duty cycle impl

adkinsjd committed:
fixed userland apps to not break with energy struct changes

bradjc commented on apps/libsignpost/signpost_api.h:
```diff
@@ -261,23 +265,39 @@ int signpost_processing_reply(uint8_t src_addr, uint8_t message_type, uint8_t* r
+//this structure is for each module
+typedef struct __attribute__((packed)) energy_report_module {
+    uint8_t module_address; //module i2c address
+    uint8_t module_percent; //an integer percent 0-100 that  the module has used
```
What is this a percent of?

adkinsjd replied:
The energy that the module used since the last update. It prevents the module (i.e. the radio) from needing to do all the math itself and it can just tell the controller who sent what percentage of packets.

bradjc replied:
Isn't that a single divide, or a subtract and divide? The math has to get done at some point, and it seems like a more clear interface might be better than move the couple lines of code from the radio to the controller.

adkinsjd replied:
Well the radio doesn't have access to the most up to date information about its own energy, the controller does. So a little math wasn't really the best explanation. It's more like ask the controller for your energy, then split it up, then report back, (possibly ensure nothing changes before you report back), then the controller applies the update (probably requiring a verification that the radio didn't say other modules used more energy than it used in the first place). I actually don't think this interface is unclear at all, especially considering that time and number of discreet events are often a great proxy for energy and lend themselves well to percent usage reporting. Once the radio module starts submetering components and this will probably have to get more complicated, but this works well for now, and would always be sufficient if the radio module only had one physical radio.

ppannuto replied:
okay I'm late to the game on this one, but I'm really confused by this interface, I feel like it's missing really important information to attribute energy to other modules

  - What's the window of time that this "energy report" should be for? Since the last call? Does the mean if I start a radio operation I have to call this once with a null module before the op, then send the packet, then again after? (Or else if I'm the first packet in the morning I get charged with all the idle power draw of the radio?)
  - It feels way more likely that the radio would turn on to send a bunch of packets, I'd have expected this to be an array of percentages for all modules (with a num_modules in packet as well for generalizability), especially given the latency to send/reply i2c messages / measure energy, plus the high energy startup/shutdown/idle costs of radios, (the costs of which you want to amortize across all the modules that demanded a wakeup)

I don't think it makes a big difference whether the energy math is done on the radio or controller side

adkinsjd replied:
The window is since the last call. In practice that doesn't matter much because the radio just sends these periodically (null or not). 

I don't really understand what you are getting at with your second point at all. What's the difference between an array of percentages for all modules and not sending a report for a module when it has used none of the radio energy? Re: amortization of wakeup - it would be great to do and I'm open to suggestions of interface improvements, but I don't think those improvements should block this pull request.

In general these interfaces are not perfect. We could think up a whole slew of improvements, but they are dead simple and I have implementations for them that work pretty well right now. Something that works for many cases is better than nothing.

ppannuto replied:
Okay.. to make sure I'm understanding this right, on the controller side, the situation would roughly be:
```
static radio_energy_last_time;
static radio_total_energy_used;
static modules_total_energy_used[8];

on_energy_report_module(addr, percent) {
    radio_energy_now = get_energy()
    energy_used = radio_energy_now - radio_energy_last_time;
   
   modules_total_energy_used[addr] += energy_used * percent
   radio_total_energy_used += energy_used * (1-percent)

    radio_energy_last_time = radio_energy_now
}
```

What's weird to me about this interface is that the windowing is implicit and it doesn't seem to map to operations very well. My intuition would be an interface of either:

  - Monitor operation API:
     - Starting operation on behalf of module X
     - Stopping operation on behalf of module X
     - All energy between last stop and next start charged to radio, all time between start and stop charged to X
  - Periodic API (closer to what you have):
     - At intervals send what %age of time has been spent on whom (60% self, 30% mod2, 10% mod4)

what's there now feels like some weird mix between these two approaches (you're sort of implicitly doing start/stop based on when you choose to send this message?) I have no intuition for how this assigns energy fairly or correctly to modules on the radio's behalf - I can see that it will charge some amount of energy to one module sometimes, so by that definition it's working, but I don't see how the energy that's getting charged is the correct amount

Maybe I'm just spectacularly confused

bradjc replied:
Is there a usecase that the "this module used Z joules" doesn't work for? I understand it would require the radio requesting its own energy periodically, but it seems like a more general interface. It would map better to the submetering case (which it sounds like is in the pipeline), and it would also map better to the case where a module is doing its own operation that should not get attributed to any module (maybe an app update over GSM for example).

I think I'm trying to understand if you are arguing that the "percent of my total" energy is very close to (or even the) interface that we want, or if it is just what exists now and it could be changed entirely in a future version. If it's the latter, then I think it might be worth having a discussion now where we can get closer to the interface we actually want so that there is less future code re-writing.

bradjc commented on apps/libsignpost/signpost_api.h:
```diff
@@ -261,23 +265,39 @@ int signpost_processing_reply(uint8_t src_addr, uint8_t message_type, uint8_t* r
+_Static_assert(sizeof(signpost_energy_information_t) == 8, "On-wire structure size");
+
+//a mechanism for modules to report energy usage from other modules
+//this structure is for each module
```
More comment would be great here. I didn't understand this until I read the PR summary.

bradjc commented on apps/libsignpost/signpost_api.h:
```diff
@@ -261,23 +265,39 @@ int signpost_processing_reply(uint8_t src_addr, uint8_t message_type, uint8_t* r
+//we make an array of them to report full usage
+typedef struct __attribute__((packed)) energy_report {
+    uint8_t num_reports;
+    signpost_energy_report_module_t reports[8];
```
Do we assume 8 modules anywhere else in the signpost API? If not, it might not be a good idea to start now.

adkinsjd replied:
ah yeah that was a hack to easily pack messages it can get removed

bradjc commented on apps/libsignpost/signpost_api.h:
```diff
@@ -261,23 +265,39 @@ int signpost_processing_reply(uint8_t src_addr, uint8_t message_type, uint8_t* r
-    uint16_t    current_limit_60s_mA;
-    uint16_t    current_average_60s_mA;
+    uint32_t    energy_limit_mWh;
+    uint16_t    average_power_mW;
```
It seems odd that the limit is specified as energy, yet the feedback is in power. How is a module expected to use this? Does its `energy_limit`decrease over time and that is how it gets its energy consumption? It could calculate how long until it exhausts its energy limit given its average power, but the `energy_information` struct doesn't include any information about how long that energy limit is valid for.

adkinsjd replied:
The module uses this by dividing limit by the their power and seeing if it is above or below some target lifetime. The energy limit is just valid. There is no timeout. It decreases if you use energy, and it increases if the signpost successfully harvests energy. We could do some windowed update but it makes everyone's logic more complex. This is simple and it works.

adkinsjd replied:
I mean this is simple and it works for now. I expect this to get more complex over time, but I think having an energy interface that is useful now is worthwhile

bradjc replied:
Ok that's reasonable. What is the suggested lifetime that a module should use? I might think, "well, we'll probably get more energy sometime over the week, so I'll set my lifetime to 7 days and see how often I can sample." I can also see, "well, my study is supposed to last 6 months, and I have no idea when energy will come in, so I'll set my lifetime to 6 months and sample at that rate." (which would likely end up with the module taking progressively more samples as the 6 month period gets closer to over). Obviously those are going to give very different results.

So while this is simple (only two values) it may not be all that simple from a module's point of view since the controller is doing no work to estimate when new energy might be available.

adkinsjd replied:
As I said I think it could get much more complex, but this allows us to write energy adaptive apps now

bradjc replied:
What do existing apps use as their lifetime?

adkinsjd replied:
~3 days

adkinsjd replied:
That audio app in the paper used a 3 day lifetime and that tracked allocated energy well.

adkinsjd replied:
That's the only app that has used it and it isn't merged yet because I'm trying to do this a bit at a time.

bradjc commented on apps/libsignpost/signpost_api.h:
```diff
@@ -261,23 +265,39 @@ int signpost_processing_reply(uint8_t src_addr, uint8_t message_type, uint8_t* r
     EnergyLevelWarning24hMessage = 1,
     EnergyLevelCritical24hMessage = 2,
     EnergyCurrentWarning60sMessage = 3,
+    EnergyReportMessage = 4,
```
I think this name could be more specific. Something like `EnergyReportModuleConsumptionMessage`.

bradjc reviewed (changes requested):
Need to update https://github.com/lab11/signpost-software/blob/master/docs/ApiGuide.md#energy as well before this can get merged.

ppannuto commented on apps/libsignpost/signpost_api.c:
```diff
@@ -856,13 +853,18 @@ static bool energy_query_ready;
     energy_query_result = result;
 }
 
+static void signpost_energy_report_callback(__attribute__ ((unused)) int result) {
```
Is it really safe to not use this result? It's quite possibly than an async error bubbles up

adkinsjd replied:
That should return a result

ppannuto commented on apps/libsignpost/signpost_api.c:
```diff
@@ -919,18 +921,60 @@ int signpost_energy_query_async(
-    if (rc < 0) return rc;
+
+    //This properly catches the error if the send fails
+    //and allows for subsequent calls to query async to succeed
```
nit: missing spaces

ppannuto commented on apps/libsignpost/signpost_api.c:
```diff
@@ -919,18 +921,60 @@ int signpost_energy_query_async(
+    int rc;
+    rc = signpost_api_send(ModuleAddressController,
+            CommandFrame, EnergyApiType, EnergyReportMessage,
+            report->num_reports*2+1, (uint8_t*)report);
```
Shouldn't `report->num_reports*2+1` be `sizeof(*report)` ?

adkinsjd replied:
not after I fix the static number of reports - but yes.

ppannuto commented on apps/libsignpost/signpost_api.c:
```diff
@@ -919,18 +921,60 @@ int signpost_energy_query_async(
+    //there is an integer in the incoming message that should be
+    //sent back as the return code
+    int return_code;
+    memcpy(&return_code,incoming_message,sizeof(int));
```
why not just
`return *incoming_message` ?

ppannuto commented on apps/libsignpost/signpost_api.h:
```diff
@@ -261,23 +265,39 @@ int signpost_processing_reply(uint8_t src_addr, uint8_t message_type, uint8_t* r
 /* ENERGY API                                                             */
 /**************************************************************************/
 
+//message types
```
I don't think this comment provides much value?

ppannuto commented on apps/libsignpost/signpost_api.h:
```diff
@@ -302,6 +334,10 @@ int signpost_energy_query_async(signpost_energy_information_t* energy, signbus_a
 __attribute__((warn_unused_result))
 int signpost_energy_query_reply(uint8_t destination_address, signpost_energy_information_t* info);
 
+//response from controller to requesting module
```
`// R`

ppannuto commented on apps/tests/api_energy_test/main.c:
```diff
@@ -37,12 +37,10 @@ int main (void) {
-      printf("  mJ limit warn: %-4u %%\n",  info.energy_limit_warning_threshold);
-      printf("  mJ limit crit: %-4u %%\n",  info.energy_limit_critical_threshold);
+      printf("      energy limit: %-4lu mWh\n", info.energy_limit_mWh);
+      printf("       average power: %-4hu mW\n", info.average_power_mW);
```
These were spaced so the `:`'s and numbers would all line up, looks like they aren't now?

adkinsjd committed:
fixed pat and some of Brad's reviews

adkinsjd committed:
missed a style nit

adkinsjd commented:
All of pat's changes are fixed and I fixed all of Brad's comments except for the ones I disagree with.

adkinsjd commented:
I still don't understand why you are confused.

Why would it not assign energy (at least close) to correctly? Assume you
are using one radio interface and the radio module sent a report 10 minutes
ago. The radio module keeps track of the modules who sent packets over the
next 10 minutes. I think we can all agree that the energy used during that
10 minutes should be split based on the percentage of packets sent by each
module in that 10 minute window. So then I send off a report assigning
percentages based on the number of packets sent and the controller reads
how much energy you have used in the last 10 minutes, splits it up from
percentages, then assigns it to the other modules.

And sure it does a poor job of splitting up idle power, but that should be
minimal compared to radio energy, and this is a first cut interface anyways
so whatever.

On Wed, Jun 7, 2017 at 12:34 PM, Pat Pannuto <notifications@github.com>
wrote:

> *@ppannuto* commented on this pull request.
> ------------------------------
>
> In apps/libsignpost/signpost_api.h
> <https://github.com/lab11/signpost-software/pull/9#discussion_r120723963>:
>
> >      uint8_t     energy_limit_warning_threshold;
>      uint8_t     energy_limit_critical_threshold;
>  } signpost_energy_information_t;
>
> -_Static_assert(sizeof(signpost_energy_information_t) == 14, "On-wire structure size");
> +_Static_assert(sizeof(signpost_energy_information_t) == 8, "On-wire structure size");
> +
> +//a mechanism for modules to report energy usage from other modules
> +//this structure is for each module
> +typedef struct __attribute__((packed)) energy_report_module {
> +    uint8_t module_address; //module i2c address
> +    uint8_t module_percent; //an integer percent 0-100 that  the module has used
>
> Okay.. to make sure I'm understanding this right, on the controller side,
> the situation would roughly be:
>
> static radio_energy_last_time;
> static radio_total_energy_used;
> static modules_total_energy_used[8];
>
> on_energy_report_module(addr, percent) {
>     radio_energy_now = get_energy()
>     energy_used = radio_energy_now - radio_energy_last_time;
>
>    modules_total_energy_used[addr] += energy_used * percent
>    radio_total_energy_used += energy_used * (1-percent)
>
>     radio_energy_last_time = radio_energy_now
> }
>
> What's weird to me about this interface is that the windowing is implicit
> and it doesn't seem to map to operations very well. My intuition would be
> an interface of either:
>
>    - Monitor operation API:
>       - Starting operation on behalf of module X
>       - Stopping operation on behalf of module X
>       - All energy between last stop and next start charged to radio, all
>       time between start and stop charged to X
>    - Periodic API (closer to what you have):
>       - At intervals send what %age of time has been spent on whom (60%
>       self, 30% mod2, 10% mod4)
>
> what's there now feels like some weird mix between these two approaches
> (you're sort of implicitly doing start/stop based on when you choose to
> send this message?) I have no intuition for how this assigns energy fairly
> or correctly to modules on the radio's behalf - I can see that it will
> charge some amount of energy to one module sometimes, so by that definition
> it's working, but I don't see how the energy that's getting charged is the
> correct amount
>
> Maybe I'm just spectacularly confused
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lab11/signpost-software/pull/9#discussion_r120723963>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGIJUBFTMiJt5ETBRarWPsT8QLwVlTJZks5sBvtFgaJpZM4NvkUw>
> .
>

adkinsjd commented:
Your periodic API is exactly what I have except (as you pointed out) the
windows are implicit. I don't understand why implicit windows make that API
any less valid.

On Wed, Jun 7, 2017 at 12:55 PM, Joshua Adkins <adkinsjd@umich.edu> wrote:

> I still don't understand why you are confused.
>
> Why would it not assign energy (at least close) to correctly? Assume you
> are using one radio interface and the radio module sent a report 10 minutes
> ago. The radio module keeps track of the modules who sent packets over the
> next 10 minutes. I think we can all agree that the energy used during that
> 10 minutes should be split based on the percentage of packets sent by each
> module in that 10 minute window. So then I send off a report assigning
> percentages based on the number of packets sent and the controller reads
> how much energy you have used in the last 10 minutes, splits it up from
> percentages, then assigns it to the other modules.
>
> And sure it does a poor job of splitting up idle power, but that should be
> minimal compared to radio energy, and this is a first cut interface anyways
> so whatever.
>
> On Wed, Jun 7, 2017 at 12:34 PM, Pat Pannuto <notifications@github.com>
> wrote:
>
>> *@ppannuto* commented on this pull request.
>> ------------------------------
>>
>> In apps/libsignpost/signpost_api.h
>> <https://github.com/lab11/signpost-software/pull/9#discussion_r120723963>
>> :
>>
>> >      uint8_t     energy_limit_warning_threshold;
>>      uint8_t     energy_limit_critical_threshold;
>>  } signpost_energy_information_t;
>>
>> -_Static_assert(sizeof(signpost_energy_information_t) == 14, "On-wire structure size");
>> +_Static_assert(sizeof(signpost_energy_information_t) == 8, "On-wire structure size");
>> +
>> +//a mechanism for modules to report energy usage from other modules
>> +//this structure is for each module
>> +typedef struct __attribute__((packed)) energy_report_module {
>> +    uint8_t module_address; //module i2c address
>> +    uint8_t module_percent; //an integer percent 0-100 that  the module has used
>>
>> Okay.. to make sure I'm understanding this right, on the controller side,
>> the situation would roughly be:
>>
>> static radio_energy_last_time;
>> static radio_total_energy_used;
>> static modules_total_energy_used[8];
>>
>> on_energy_report_module(addr, percent) {
>>     radio_energy_now = get_energy()
>>     energy_used = radio_energy_now - radio_energy_last_time;
>>
>>    modules_total_energy_used[addr] += energy_used * percent
>>    radio_total_energy_used += energy_used * (1-percent)
>>
>>     radio_energy_last_time = radio_energy_now
>> }
>>
>> What's weird to me about this interface is that the windowing is implicit
>> and it doesn't seem to map to operations very well. My intuition would be
>> an interface of either:
>>
>>    - Monitor operation API:
>>       - Starting operation on behalf of module X
>>       - Stopping operation on behalf of module X
>>       - All energy between last stop and next start charged to radio,
>>       all time between start and stop charged to X
>>    - Periodic API (closer to what you have):
>>       - At intervals send what %age of time has been spent on whom (60%
>>       self, 30% mod2, 10% mod4)
>>
>> what's there now feels like some weird mix between these two approaches
>> (you're sort of implicitly doing start/stop based on when you choose to
>> send this message?) I have no intuition for how this assigns energy fairly
>> or correctly to modules on the radio's behalf - I can see that it will
>> charge some amount of energy to one module sometimes, so by that definition
>> it's working, but I don't see how the energy that's getting charged is the
>> correct amount
>>
>> Maybe I'm just spectacularly confused
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/lab11/signpost-software/pull/9#discussion_r120723963>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AGIJUBFTMiJt5ETBRarWPsT8QLwVlTJZks5sBvtFgaJpZM4NvkUw>
>> .
>>
>
>

adkinsjd committed:
added implementation of the energy policy described in energy api

adkinsjd committed:
adds energy reporting, sequence number, turns off manual reset checks

We should probably turn back on manual reset checks but they were
not working. Should also update to xdot seeing as this will not work
on new radio module

adkinsjd committed:
energy send bytes app used for Sensys

Changes:
 - implements duty cycling
 - implements energy policy including module cuteoff and energy policy init from FRAM
 - Changes information sent in lora packets to include bat/solar info
 - implements energy reporting

adkinsjd commented:
Okay I have (finally) integrated the files that depend on any changes to the energy API - this should allow us to move forward with 1) changing the API to be more expressive as necessary and 2) generally cleaning up the PR.

I am going to put a summary of the work that I think needs to be done to get this merged in my initial comment at the top of the PR.

adkinsjd commented:
I may also pull in the audio and ambient apps that use adaptation just because they are small and show how the current API was used in the paper

adkinsjd committed:
added the adaptive audio app used in the paper

adkinsjd committed:
changed the ambient sense_and_send app to be duty cycling (as in paper)

adkinsjd committed:
updated energy api, started implementation of update in policy file

adkinsjd committed:
fixed error codes in energy API

adkinsjd committed:
implemented new energy api functions

adkinsjd committed:
refactored out controller features into their own file

this included making many of the function asynchronous
and also adding some code to the energy policy file

adkinsjd committed:
consolidated module state into struct

adkinsjd committed:
added energy reset impl

adkinsjd committed:
fixed ambient adaptive to new interface

adkinsjd committed:
fixed energy api test

adkinsjd committed:
made this app compile

This should really be deprecated, but we still use it on debug backplanes??

adkinsjd committed:
made app compile

adkinsjd committed:
fixed a timer, added print statements

adkinsjd committed:
got this branch back up to date with master

adkinsjd committed:
added simplest controller app possible

adkinsjd committed:
Removed unnecessary bonus timer
